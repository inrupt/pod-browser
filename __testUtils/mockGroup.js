/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  getStringNoLocale,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/group";
import { vcardExtras } from "../src/models/addressBook";

function mockGroupThing(name, groupThingUrl) {
  return chain(
    mockThingFrom(groupThingUrl),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, vcard.fn, name)
  );
}

function mockIndexThing(addressBook, groupThingUrl) {
  return setUrl(addressBook.thing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
  );
}

// export function addGroupsToAddressBook(addressBook, groupsToAdd) {
//   const { groups, ...rest } = addressBook;
//   return {
//     ...rest,
//     groups: {
//       iri: addressBook.groups.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...groupsToAdd.map(({ dataset, iri }) => (d) => {
//           const group = getThing(dataset, iri);
//           const name = getStringNoLocale(group, vcard.fn);
//           return setThing(d, mockGroupThing(name, iri));
//         }),
//         ...groupsToAdd.map((group) => (d) =>
//           setThing(d, mockIndexThing(addressBook, group.iri))
//         )
//       ),
//     },
//   };
// }

export default function mockGroup(
  addressBook,
  name,
  { url, id, members } = {}
) {
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = chain(
    mockGroupThing(name, groupThingUrl),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    )
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
