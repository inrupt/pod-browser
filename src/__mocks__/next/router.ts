export default {
  push: jest.fn().mockResolvedValue(null),
};
