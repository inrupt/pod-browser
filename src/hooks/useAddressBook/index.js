/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useContext, useEffect, useState } from "react";
import { fetchLitDataset } from "@inrupt/solid-client";
import SessionContext from "../../contexts/sessionContext";
import useAuthenticatedProfile from "../useAuthenticatedProfile";
import { contactsContainerIri, saveNewAddressBook } from "../../addressBook";
import { getResource } from "../../solidClientHelpers/resource";
import { ERROR_CODES, isHTTPError } from "../../solidClientHelpers/utils";

export default function useAddressBook() {
  const [addressBook, setAddressBook] = useState(null);
  const [error, setError] = useState(null);
  const { session } = useContext(SessionContext);
  const profile = useAuthenticatedProfile();

  useEffect(() => {
    if (!session.info.isLoggedIn || !profile) return;
    const { pods, webId } = profile;
    const { fetch } = session;
    const contactsIri = contactsContainerIri(pods[0]);

    (async () => {
      let addressBookUri;

      const {
        response: existingAddressBook,
        error: existingError,
      } = await getResource(contactsIri, fetch);

      if (existingAddressBook) {
        setAddressBook(existingAddressBook.dataset);
        return;
      }

      if (existingError && !isHTTPError(existingError, ERROR_CODES.NOT_FOUND)) {
        const {
          response: newAddressBook,
          error: newError,
        } = await saveNewAddressBook(
          {
            iri: contactsIri,
            owner: webId,
          },
          fetch
        );
        if (newError) {
          setError(newError);
          return;
        }
        addressBookUri = newAddressBook.iri;
      } else if (existingError) {
        setError(existingError);
        return;
      }

      setAddressBook(await fetchLitDataset(addressBookUri, { fetch }));
    })();
  }, [session, profile]);

  return [addressBook, error];
}
