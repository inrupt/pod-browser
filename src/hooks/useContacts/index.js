/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useSession } from "@inrupt/solid-ui-react";
import useSWR from "swr";
import { thingAsMarkdown } from "@inrupt/solid-client";
import useAddressBook from "../useAddressBook";
import { getContactAll } from "../../models/contact";
import { compareArray } from "../../models/array";

function reduceToString(array) {
  return array.reduce(
    (memo, item) => memo.concat(thingAsMarkdown(item.thing)),
    ""
  );
}

export default function useContacts(type) {
  const { fetch } = useSession();
  const { data: addressBook, error: addressBookError } = useAddressBook();
  return useSWR(
    ["contacts", addressBook, type],
    async () => {
      if (!addressBook && !addressBookError) return null;
      if (addressBookError) throw addressBookError;
      return getContactAll(addressBook, [type], fetch);
    },
    {
      // revalidateOnMount: true,
      // revalidateOnFocus: false,
      // revalidateOnReconnect: false,
      // refreshInterval: 0,
      shouldRetryOnError: false,
      // errorRetryCount: 0,
      compare: (a, b) =>
        compareArray(a, b, (c, d) => reduceToString(c) === reduceToString(d)),
      // compare: (a, b) => {
      //   if (typeof a === "undefined") {
      //     return typeof b === "undefined";
      //   }
      //   if (typeof b === "undefined") {
      //     return typeof a === "undefined";
      //   }
      //   if (!Array.isArray(a) || !Array.isArray(b)) {
      //     return false;
      //   }
      //   if (a.length !== b.length) {
      //     return false;
      //   }
      //   const reduce = reduceToString(a) === reduceToString(b);
      //   // if (!reduce) {
      //   //   console.log("COMPARE ARRAY A", reduceToString(a));
      //   //   console.log("COMPARE ARRAY B", reduceToString(b));
      //   // }
      //   return reduce;
      // },
    }
  );
}
