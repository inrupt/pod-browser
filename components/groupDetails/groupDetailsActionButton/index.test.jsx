/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { useRouter } from "next/router";
import userEvent from "@testing-library/user-event";
import renderGroupsPage from "../../../__testUtils/renderGroupsPage";
import useAddressBook from "../../../src/hooks/useAddressBook";
import useContacts from "../../../src/hooks/useContacts";
import useGroup from "../../../src/hooks/useGroup";
import mockGroup from "../../../__testUtils/mockGroup";
import {
  TESTCAFE_ID_GROUP_DETAILS_MODAL,
  TESTCAFE_ID_GROUP_DETAILS_MODAL_CANCEL_BUTTON,
} from "../groupDetailsModal";
import GroupDetailsActionButton, {
  TESTCAFE_ID_GROUP_DETAILS_ACTION_BUTTON_DELETE_ACTION,
  TESTCAFE_ID_GROUP_DETAILS_ACTION_BUTTON_EDIT_ACTION,
} from "./index";
import {
  TESTCAFE_ID_GROUP_DELETE_MODAL,
  TESTCAFE_ID_GROUP_DELETE_MODAL_CANCEL_BUTTON,
} from "../groupDeleteModal";

jest.mock("../../../src/hooks/useAddressBook");
const mockedAddressBookHook = useAddressBook;

jest.mock("../../../src/hooks/useContacts");
const mockedContactsHook = useContacts;

jest.mock("../../../src/hooks/useGroup");
const mockedGroupHook = useGroup;

jest.mock("next/router");
const mockedRouterHook = useRouter;

const group1Name = "Group 1";
const group1Url = "http://example.com/group1.ttl#this";
const group1 = mockGroup(group1Name, group1Url);

describe("GroupDetailsActionButton", () => {
  beforeEach(() => {
    mockedAddressBookHook.mockReturnValue({});
    mockedContactsHook.mockReturnValue({ data: [group1] });
    mockedGroupHook.mockReturnValue({ data: group1 });
    mockedRouterHook.mockReturnValue({ query: { iri: group1Url } });
  });

  it("renders", () => {
    const { asFragment, getByTestId } = renderGroupsPage(
      <GroupDetailsActionButton />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(
      getByTestId(TESTCAFE_ID_GROUP_DETAILS_ACTION_BUTTON_EDIT_ACTION)
    ).toBeDefined();
  });

  it("opens GroupDetailsModal", () => {
    const { getByTestId, queryByTestId } = renderGroupsPage(
      <GroupDetailsActionButton />
    );
    expect(queryByTestId(TESTCAFE_ID_GROUP_DETAILS_MODAL)).toBeNull();
    userEvent.click(
      getByTestId(TESTCAFE_ID_GROUP_DETAILS_ACTION_BUTTON_EDIT_ACTION)
    );
    expect(getByTestId(TESTCAFE_ID_GROUP_DETAILS_MODAL)).toBeDefined();
    userEvent.click(getByTestId(TESTCAFE_ID_GROUP_DETAILS_MODAL_CANCEL_BUTTON));
    expect(queryByTestId(TESTCAFE_ID_GROUP_DETAILS_MODAL)).toBeNull();
  });

  it("opens GroupDeleteModal", () => {
    const { getByTestId, queryByTestId } = renderGroupsPage(
      <GroupDetailsActionButton />
    );
    expect(queryByTestId(TESTCAFE_ID_GROUP_DELETE_MODAL)).toBeNull();
    userEvent.click(
      getByTestId(TESTCAFE_ID_GROUP_DETAILS_ACTION_BUTTON_DELETE_ACTION)
    );
    expect(getByTestId(TESTCAFE_ID_GROUP_DELETE_MODAL)).toBeDefined();
    userEvent.click(getByTestId(TESTCAFE_ID_GROUP_DELETE_MODAL_CANCEL_BUTTON));
    expect(queryByTestId(TESTCAFE_ID_GROUP_DELETE_MODAL)).toBeNull();
  });
});
