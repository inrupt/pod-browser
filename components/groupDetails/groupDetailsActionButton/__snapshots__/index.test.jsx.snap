// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupDetailsActionButton renders 1`] = `
<DocumentFragment>
  <div>
    <button
      class="PodBrowser-button PodBrowser-button--icon-only PodBrowser-action-button"
    >
      <span
        class="PodBrowser-button__text"
      >
        <span
          aria-label="Show menu"
          class="PodBrowser-icon-more"
        />
      </span>
    </button>
  </div>
</DocumentFragment>
`;
